src/host/coreaudio/macos/property_listener.rs

```
@@ -0,0 +1,85 @@
+/// The listener will be removed when this type is dropped.
+pub struct AudioObjectPropertyListener {
+    callback: Box<PropertyListenerCallbackWrapper>,
+    property_address: AudioObjectPropertyAddress,
+    audio_object_id: AudioObjectID,
+    removed: bool,
+}
+
+impl AudioObjectPropertyListener {
+    /// Attach the provieded callback as a audio object property listener.
+    pub fn new<F: FnMut() -> () + 'static>(
+        audio_object_id: AudioObjectID,
+        property_address: AudioObjectPropertyAddress,
+        callback: F,
+    ) -> Result<Self, BuildStreamError> {
+        let callback = Box::new(PropertyListenerCallbackWrapper(Box::new(callback)));
+        unsafe {
+            coreaudio::Error::from_os_status(AudioObjectAddPropertyListener(
+                audio_object_id,
+                &property_address as *const _,
+                Some(property_listener_handler_shim),
+                &*callback as *const _ as *mut _,
+            ))?;
+        };
+        Ok(Self {
+            callback,
+            audio_object_id,
+            property_address,
+            removed: false,
+        })
+    }
+
+    /// Explicitly remove the property listener.
+    /// Use this method if you need to explicitly handle failure to remove
+    /// the property listener.
+    pub fn remove(mut self) -> Result<(), BuildStreamError> {
+        self._remove()
+    }
+
+    fn _remove(&mut self) -> Result<(), BuildStreamError> {
+        unsafe {
+            coreaudio::Error::from_os_status(AudioObjectRemovePropertyListener(
+                self.audio_object_id,
+                &self.property_address as *const _,
+                Some(property_listener_handler_shim),
+                &*self.callback as *const _ as *mut _,
+            ))?;
+        }
+        self.removed = true;
+        Ok(())
+    }
+}
+
+impl Drop for AudioObjectPropertyListener {
+    fn drop(&mut self) {
+        if !self.removed {
+            self._remove().ok();
```

I think `_ = ...` (or `let _ = ...`) is better than calling ok here, to ignore the error.

---

Heh, I'm so accustomed to "ensure every error is handled or returned" that I don't have well-formed opinions on how to ignore them 😄